ddtrace/internal/ci_visibility/_api_client.py

```diff
@@ -272,11 +286,57 @@ def _do_request(self, method: str, endpoint: str, payload: str, timeout: t.Optio
             if conn is not None:
                 conn.close()
 
+    def _do_request_with_telemetry(
+        self,
+        method: str,
+        endpoint: str,
+        payload: str,
+        metric_names: APIRequestMetricNames,
+        timeout: t.Optional[float] = None,
+    ) -> t.Any:
+        """Performs a request with telemetry submitted according to given names"""
+        sw = StopWatch()
+        sw.start()
+        error_type: t.Optional[ERROR_TYPES] = None
+        response_bytes: t.Optional[int] = None
+        try:
+            try:
+                response = self._do_request(method, endpoint, payload, timeout=timeout)
+            except _NETWORK_ERRORS:
+                error_type = ERROR_TYPES.TIMEOUT
+                raise
+            if response.status >= 400:
+                error_type = ERROR_TYPES.CODE_4XX if response.status < 500 else ERROR_TYPES.CODE_5XX
+                if response.status == 403:
+                    raise CIVisibilityAuthenticationException()
+                raise ValueError("API response status code: %d", response.status)
+            try:
+                sw.stop()  # Stop the timer before parsing the response
+                response_bytes = len(response.body)
+                if isinstance(response.body, bytes):
+                    parsed = json.loads(response.body.decode())
+                else:
+                    parsed = json.loads(response.body)
+                return parsed
+            except JSONDecodeError:
+                error_type = ERROR_TYPES.BAD_JSON
+                raise
+        finally:
+            record_api_request(metric_names, sw.elapsed() * 1000, response_bytes=response_bytes, error=error_type)
```

Comment:
...ah, we record it here now, right? So recording the request is separate from recording whether there were errors.

Reply:
Yes, all the metrics are technically added separately and there's no association between them.

The only reason they were all in a single function call was because it is (or seems to be, at least) more logical.